### ISSUE TITLE
[Feature Request] Reduce ops with keepdim=False are not supported

### ISSUE THREAD

**Issue by mrakitaTT** [tenstorrent/tt-metal#13361]
Currently ttnn reduce ops throw error when called with `keepdim=False`. There is a check inside `tt-metal/ttnn/cpp/ttnn/operations/reduction/generic/generic_reductions.cpp` in `reduce_impl` function which checks for this and throws error.

This presents an issue for us in tt-mlir compiler because then we have to always manually reshape the returned tensor after calling reduce ops with `keepdim=True`.

Do you have any plans to support this feature in the near future?

**Comment by tt-mpantic** [comment]
P0 since it is blocking JAX MNIST.

**Comment by mrakitaTT** [comment]
> Do you have any plans to support this feature in the near future?

Hi @ntarafdar, could you please provide some info on this?

**Comment by ntarafdar** [comment]
Hey @mrakitaTT , can you ask @bbradelTT . Currently reduction has no home but I think Jasmina mentioned he might be able to take them.

**Comment by mrakitaTT** [comment]
> Do you have any plans to support this feature in the near future?

Hi @bbradelTT, could you please provide some info on this? :)

**Comment by bbradelTT** [comment]
I'll take a look.

**Comment by bbradelTT** [comment]
@mrakitaTT can you please provide all of the calls that you are interested in (including shapes if input tensors, etc.)?

With recent changes, depending on the shape, sometimes there are downstream fatals, sometimes there aren't. Depending on the shapes you are interested in, it may be possible to unblock you sooner.

**Comment by mrakitaTT** [comment]
Reducing to P1 since even though it is blocking MNIST model we have a way to make a workaround for this issue in compiler if this will take a lot of time to fix on Metal side.

**Comment by mrakitaTT** [comment]
@bbradelTT Sure, here are the traces for MNIST model:

| Reduce op | dimArg | Input tensor shape | Output tensor shape |
| --- | --- | --- | --- |
| ttnn::max | 1 | tensor<128x10xf32> | tensor<128xf32> |
| ttnn::sum | 1 | tensor<128x10xf32> | tensor<128xf32> |
| ttnn::sum | 1 | tensor<128x1xf32> | tensor<128xf32> |
| ttnn::sum | 0 | tensor<128x10xf32> | tensor<10xf32> |
| ttnn::sum | 0 | tensor<1x10xf32> | tensor<10xf32> |
| ttnn::sum | 0 | tensor<128x512xf32> | tensor<512xf32> |
| ttnn::sum | 0 | tensor<1x512xf32> | tensor<512xf32> |

**Comment by bbradelTT** [comment]
I'll try out the shapes tomorrow. From my own investigation, it looks like the problem is in reduce, which seems to be called when keepdim=False.

```
import torch
import ttnn
device_id = 0
device = ttnn.open_device(device_id=device_id)
e=ttnn.from_torch(torch.ones([1,1,32,1], dtype=torch.bfloat16), layout=ttnn.TILE_LAYOUT, device=device)
s=ttnn.Shape([32],[32])
f=ttnn.reshape(e,s)
ttnn.close_device(device)
exit()
```

results in

```
RuntimeError: TT_FATAL @ ../ttnn/cpp/ttnn/tensor/types.cpp:211: normalized_index >= 0 and normalized_index < rank
info:
Index is out of bounds for the rank, should be between 0 and 0 however is 18446744073709551615
backtrace:
 --- /localdev/bbradel/tt-metal/ttnn/ttnn/_ttnn.so(+0x1a5c7cb) [0x7f05fab907cb]
 --- tt::tt_metal::LegacyShape::get_normalized_index(long) const
 --- tt::tt_metal::LegacyShape::operator[](long) const
 --- ttnn::types::Shape::operator[](long) const
 --- ttnn::operations::data_movement::ReshapeViewOperation::invoke(tt::tt_metal::Tensor const&, ttnn::types::Shape const&)
...
```

I need to figure out a way to avoid this error.

**Comment by bbradelTT** [comment]
Unfortunately there are a lot of issues with padding.

E.g. some internal shapes that are showing up in my tests are:
```
[128, 1[32], 10[32], 32]
[128, 1, 10[32], 32]
[1[32], 128, 10[32], 32]
```
The new tensor shape work should make all of the padding irrelevant. We'll have to wait for that to be done, and then I'll revisit this issue.

cc @TT-BrianLiu @ayerofieiev-tt

**Comment by bbradelTT** [comment]
@mrakitaTT could you please test with newest tt-metal main and let me know what you find?

You may need to test with a release build to avoid some asserts.

**Comment by mmanzoorTT** [comment]
@bbradelTT Thanks for the patch. It solves the issue except in case of reducing for entire tensor. If we want to apply reduction for entire tensor with rank 4 (e.g. ttnn.sum(input)); the output will have a shape of `1x1x1x1xf32`; which is inconsistent with pytorch or other packages. The expected shape is `1xf32`

**Comment by bbradelTT** [comment]
@mmanzoorTT I'm glad to hear the other cases are solved. I need to wait until https://github.com/tenstorrent/tt-metal/issues/15416 is fixed before looking at the `1xf32` case.

**Comment by sdjordjevicTT** [comment]
Hi @bbradelTT, do we have any updates here regarding this 1xf32 case?

**Comment by bbradelTT** [comment]
@sdjordjevicTT this has been fixed for a while. The output tensor should have `shape=Shape([])`. Sorry, this issue had no status on my dashboard and I forgot to update it.

Please verify if your scenario is satisfied and either close the issue or provide a repro.

**Comment by sdjordjevicTT** [comment]
@bbradelTT no problem, thanks for the fix! We will check on our end.

**Comment by fahmedchanTT** [comment]
@sdjordjevicTT Can this be closed? Thanks

CC: @skhatriTT

**Comment by sdjordjevicTT** [comment]
Yes, we tested it on our end, and everything works fine. Thanks for the fix!
